File: code/src/main/java/service/ServiceTicket.java
package service;

import model.Performance;
import model.Ticket;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.interfaces.RepositoryInterfaceTicket;
import validator.Validator;

import java.time.LocalDateTime;
import java.util.Optional;

public class ServiceTicket {

    private RepositoryInterfaceTicket repository;
    private Validator<Ticket> validatorTicket;
    private Validator<Performance> validatorPerformance;
    private static final Logger logger = LogManager.getLogger();

    public ServiceTicket(RepositoryInterfaceTicket repository, Validator<Ticket> validatorTicket, Validator<Performance> validatorPerformance) {
        logger.info("Initializing ServiceTicket.");
        this.repository = repository;
        this.validatorTicket = validatorTicket;
        this.validatorPerformance = validatorPerformance;
    }

    public Optional<Ticket> save(Long performanceID, LocalDateTime date, String place, int noOfAvailableSeats, int noOfSoldSeats, String artist, String owner, int noOfSeats) {
        logger.traceEntry("Saving ticket at performance with ID {}, bought by {}.", performanceID, owner);
        Performance performance = new Performance(performanceID, date, place, noOfAvailableSeats, noOfSoldSeats, artist);
        validatorPerformance.validate(performance);
        Ticket ticket = new Ticket(performance, owner, noOfSeats);
        validatorTicket.validate(ticket);
        return repository.save(ticket);
    }
}

File: MusicFestivalMPP/service/ServiceTicket.cs
using System;
using log4net;
using MusicFestivalMPP.model;
using MusicFestivalMPP.repository.interfaces;
using MusicFestivalMPP.validator;

namespace MusicFestivalMPP.service;

public class ServiceTicket
{
    private IRepositoryTicket Repository;
    private IValidator<Ticket> ValidatorTicket;
    private IValidator<Performance> ValidatorPerformance;
    private static readonly ILog Log = LogManager.GetLogger("ServiceTicket");

    public ServiceTicket(IRepositoryTicket repository, IValidator<Ticket> validatorTicket, IValidator<Performance> validatorPerformance)
    {
        Log.InfoFormat("Initializing ServiceTicket.");
        Repository = repository;
        ValidatorTicket = validatorTicket;
        ValidatorPerformance = validatorPerformance;
    }

    public Ticket Save(long performanceId, DateTime date, String place, int noOfAvailableSeats, int noOfSoldSeats, String artist, String owner, int noOfSeats) {
        Log.InfoFormat("Saving ticket at performance with ID {0}, bought by {1}.", performanceId, owner);
        Performance performance = new Performance(performanceId, date, place, noOfAvailableSeats, noOfSoldSeats, artist);
        ValidatorPerformance.Validate(performance);
        Ticket ticket = new Ticket(performance, owner, noOfSeats);
        ValidatorTicket.Validate(ticket);
        return Repository.Save(ticket);
    }
}